### PR TITLE
[Bug-12960][Master] Fix that start parameters cannot update global variables

### PR DESCRIPTION
**Pull request by Radeity**
## Purpose of the pull request
- This PR will close https://github.com/apache/dolphinscheduler/issues/12960

## Brief change log
- Target version: 2.0.x
- Simply synchronize change from https://github.com/apache/dolphinscheduler/pull/9952 and https://github.com/apache/dolphinscheduler/pull/9730.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by Radeity** [comment]
@JinyLeeChina PTAL.

**Comment by brave-lee** [comment]
OK, I will test it

**Comment by brave-lee** [comment]
image
image
I tested your code but didn't get the expected results. Please confirm

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java [hunk 1]
```diff
@@ -684,6 +685,9 @@ private void initTaskQueue() {
         if (processInstance.isComplementData() && complementListDate.size() == 0) {
             Map<String, String> cmdParam = JSONUtils.toMap(processInstance.getCommandParam());
             if (cmdParam != null && cmdParam.containsKey(CMDPARAM_COMPLEMENT_DATA_START_DATE)) {
```
Comment (brave-lee) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java:690]: This is not just a complement. There should be a similar problem when you directly click Start

**Comment by Radeity** [comment]
> I tested your code but didn't get the expected results. Please confirm

I'll check later today.

**Comment by Radeity** [comment]
> image
> 
> image

Hi, @JinyLeeChina, i've checked and find nothing wrong, i wonder what's wrong in the picture? It seems like startup parameters update global parameters.

**Comment by brave-lee** [comment]
> > image
> > image
> 
> Hi, @JinyLeeChina, i've checked and find nothing wrong, i wonder what's wrong in the picture? It seems like startup parameters update global parameters.

Yes, I find `startup parameters` can update global parameters, but can't update local parameters, please check

**Comment by Radeity** [comment]
> Yes, I find `startup parameters` can update global parameters, but can't update local parameters, please check

According to the [doc](https://dolphinscheduler.apache.org/#/zh-cn/docs/3.1.2/guide/parameter/priority), local parameter have higher priority than global and startup parameters.

**Comment by Radeity** [comment]
> Yes, I find `startup parameters` can update global parameters, but can't update local parameters, please check

According to the [doc](https://dolphinscheduler.apache.org/#/zh-cn/docs/3.1.2/guide/parameter/priority), local parameter has higher priority than global and startup parameter, if you don't set local parameter with same name, then you can use different startup parameter in the task.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java (hunk 1 above)
Comment (Radeity) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java:690]: > This is not just a complement. There should be a similar problem when you directly click Start

This part fixes that cmd param can overwrite global param when executing complement, change in ProcessService fix bug when directly click Start.

Comment (brave-lee) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteThread.java:690]: I mean, it's not just a matter of complement. Can you help fix it

**Comment by brave-lee** [comment]
Good job

**Review by brave-lee — APPROVED** [review]
+1

**Pull request merged**